### ISSUE TITLE
Replace directives with annotations for relationships

### ISSUE THREAD

**Issue by craicoverflow** [aerogear/graphback#539]
The new migration library uses custom annotations (using [graphql-annotations](https://github.com/Akryum/graphql-annotations)) for many-to-many relationships. Example:

```gql
type User {
  id: ID!
  name: String
  """
  @db.manyToMany: 'friends'
  """
  friends: [User]
}

```

Graphback uses directives to map and generate relationships:

```gql
type Note {
  comment: [Comment!]! @OneToMany(field: "yourCustomField")
}
```

We should pick one common strategy for relationship implementation between graphql-migrations and Graphback.

**Comment by craicoverflow** [comment]
As per the suggestion of @wtrocki this could be extracted into a separate package for reading and abstracting annotations, which could be used by both.

**Comment by craicoverflow** [comment]
As per the suggestion of @wtrocki this could be extracted into a separate package for reading and abstracting annotations, which could be used by both.

**Comment by wtrocki** [comment]
Directives are challenging as they need to be dynamically suppied to every model in order to properly parse it. Then we do not know what kind of directives each individual element in ecosystem will require which requires us to have global list of directives. 
This on the other hand will restrict community for building their own plugins (and directives) as there will be no easy way to append them - like in graphql-code-generator. 

I think that graphql annotations perfectly resolve this issue. 
You can suddenly drop some extra elements into the commend and supply your own handler for those elements without worring if schema is valid. Bonus point is that this comments 

Directives will work for the cases where we want to actually leave the directive in generated code. For example hasRole, subscribe directive can be introduced etc.

FYI @darahayes

**Comment by darahayes** [comment]
> Directives will work for the cases where we want to actually leave the directive in generated code. For example hasRole, subscribe directive can be introduced etc.

This is a really strong point. How to safely distinguish between directives used for generation purposes which will possibly be removed from the final schema and directives that modify runtime behaviour and should be kept in the schema.

> You can suddenly drop some extra elements into the comment and supply your own handler for those elements without worrying if the schema is valid

This is a very strong point too. The annotations are a bit more portable in the sense that if they don't cause any problems if there's nothing (during generation or runtime phases) actually doing anything with them.

I'm obviously not an expert here but I think having a clear distinction between the two is important. Annotations for generation/compilation/build behaviours and directives for runtime behaviours.

There's a small risk that having a mixture of both directives and annotations could introduce a lot of syntax that the user has to understand, but so long as there is a clear usage for both, I think it could work.

**Comment by craicoverflow** [comment]
> I'm obviously not an expert here but I think having a clear distinction between the two is important. Annotations for generation/compilation/build behaviours and directives for runtime behaviours.

Agreed. Right now directives are only used for generation purposes so it should be easy to make the distinction and switch.

**Comment by craicoverflow** [comment]
> Annotations for generation/compilation/build behaviours and directives for runtime behaviours.

Another clear distinction could be: annotations for database behaviours and directives for everything else.

With this, if a developer is happy with the default database settings and doesn't want to migrate then they only need to work with directives.

Another decision we need to make is whether to use the generated schema or the the input model as the GraphQL schema source during database migration.

### Input Model

- If we use the input model then the migration library would need to be aware of non-database related directives. This is bad and I don't see this as a viable option.

### Generated schema

If we use the generated schema then we would need to transfer all annotations from the input model to the generated schema. This is less bad.

## Annotations only

Another option would be to go all in on annotations and to use the input model for migrations. This has the extra benefit of the user not needing to remember if they are to use directives or annotations for a specific behaviour.

We could use the namespace to dictate whether an annotation should be in the generated schema.

`@db.*` - don't bring across database annotations.
`@gen.*` - don't bring across.
- `@runtime.*` - bring across.

**Comment by wtrocki** [comment]
> Another clear distinction could be: annotations for database behaviours and directives for everything else.

Not possible as annotations will drive a way we query data map field names etc.

**Comment by craicoverflow** [comment]
> There is a simple way to filter out directives in model so we can do that before sending schema to db migration.

If we are going to add annotations to the generated schema anyway why not just send the generated schema to db migration?

**Comment by wtrocki** [comment]
Conceptually - it should be no difference for the migration package, but this is simply more hacky as what we see now is the model and there is no straight connection to generation of the schema.

Schema can be generated into multiple formats that not necesairly play along with the string format of the migration package. I think we focusing a lot here on runtime use case where we can  get schema, but in cli use case we always get model and this will require refactoring of the cli which is currently in maintenance mode. What I think is just to keep it light and simple for the moment

**Comment by wtrocki** [comment]
TL;DR we might need to hack around passing whole schema, while model is just there. 
For the migration package there should be no difference what we passing as Queries/Mutations do not affect db migration and we assumed we no longer do add things to schema

**Comment by craicoverflow** [comment]
> Conceptually - it should be no difference for the migration package, but this is simply more hacky as what we see now is the model and there is no straight connection to generation of the schema.

Got you.

1. Add `@db.*` annotations to `InputModel[Field/Type]Context`.
2. Replace relationship directives with `db` annotations. It should be close to a like for like swap.
3. Include `@db` annotations in generated schema.
4. Pass model schema to migration library.
5. Filter out all directives from schema from migration library. this could be added as an option to `MigrationOptions`.

**Comment by wtrocki** [comment]
100% on everything:

> Add @db.* annotations to InputModel[Field/Type]Context.

I'm assuming that we will simply build abstract relationship model based on annotation - currently we using directives everywhere which is very bad in terms of extensibility. 

We have done some of that work already - just need to follow up. 
Le'ts also remove docs that cover this case.

**Comment by wtrocki** [comment]
> Include @db annotations in generated schema.

Maybe this makes no sense now

**Comment by craicoverflow** [comment]
> Maybe this makes no sense now.

It depends on what we use it for. For now I guess we don't have any need for it.

**Comment by craicoverflow** [comment]
Closed https://github.com/aerogear/graphback/pull/562
